### ISSUE TITLE
imgproxy: autoquality (aq) — iterative encode quality search

### ISSUE THREAD

**Issue by hlindset** [hlindset/image_plug#344]
## `autoquality` (`aq`) — iterative encode quality search

imgproxy **Pro** option. ⭕ Missing (`docs/imgproxy_support_matrix.md` → "Output and encoding").

```
autoquality:size:%target:%min_quality:%max_quality
aq:size:%target:%min_quality:%max_quality

autoquality:dssim:%target:%min_quality:%max_quality:%allowed_error
aq:dssim:%target:%min_quality:%max_quality:%allowed_error
```

Selects encode quality by searching toward a target (by size or a quality metric like DSSIM, depending on `method`) within `[min_quality, max_quality]`. The grammar is **per-method**: `allowed_error` is only present for `dssim` (and `ml`); the `size` form takes no `allowed_error`. `autoquality:none` disables it.

> Note: `autoquality` is entirely Pro/closed in imgproxy — it is **not** in the open-source option keys (only plain `quality`/`format_quality` are), so the grammar above is sourced from the [docs examples](https://docs.imgproxy.net/features/autoquality), not the upstream parser.

### Method scope

Support `none`, `size`, `dssim`. **Defer `ml`** — it predicts the initial quality from a trained ONNX net (256×256×3 RGB → 1 scalar), but there is no license-clean prebuilt model: imgproxy's ship only in the proprietary Pro `-ml` Docker image and the training methodology is unpublished. Reject `ml` (and any other unknown method) before side effects, and document it as a deliberate divergence. The open alternative, if ever pursued, is metric-targeted search (SSIMULACRA2/butteraugli), not a quality-prediction CNN. See [research note](https://github.com/hlindset/image_pipe/issues/344#issuecomment-4752643993).

Per-format quality clamps (`IMGPROXY_AUTOQUALITY_FORMAT_MIN`/`FORMAT_MAX`, e.g. `avif=65`) are **config-only** in imgproxy — there is no processing-option equivalent. The parser row covers the processing-option form above; format clamps, if implemented, belong on the config surface.

### Relationship to `max_bytes` (#83)

This is the sibling of `max_bytes`/`mb` (#83 — "Support imgproxy max_bytes output-size targets"): both are encode-search loops at the output boundary that re-encode at varying quality until a target is met. They should share the same encode-search machinery — worth designing them together even if they land separately. File this as the `autoquality` half; `max_bytes` stays in #83.

### Realization notes

- Lives at the output/encoder boundary (`lib/image_pipe/output/encoder.ex`) as a re-encode loop, not a transform op.
- The search encodes the **full-res result**. A downsized proxy may only *seed* the initial bracket — never be the sole search target — because neither size nor DSSIM is scale-invariant (the DSSIM bias is even non-monotonic in scale). See [research note](https://github.com/hlindset/image_pipe/issues/344#issuecomment-4752643993).
- For `size` specifically, **ML is not required**: the quality→size curve is smooth and monotonic per image, so 1–2 probe encodes + interpolation can hit the target within tolerance. Evaluate whether a couple of probes is cheap enough before reaching for anything fancier.
- Methods beyond byte-size (perceptual metrics) may need extra deps or a libvips capability check; scope which `method` values are supported and reject the rest before side effects.
- Pro option → acceptance via wire/header tests (resulting size within tolerance), not a pixel differential.

### Definition of done

- Parser row + `Plan.Output` field(s) + encode-search loop.
- Fiddle UI controls + URL state.
- `docs/imgproxy_support_matrix.md` updated (surface axis); document unsupported `method` values (`ml`) as a deliberate divergence.

**Comment by hlindset** [comment]
## Research note: downsized-proxy quality search + the ML/metric landscape

Scoping research for two questions: (1) can the encode-quality search run on a **downsized proxy** for speed without losing accuracy? (2) are there **prebuilt ML models / published methods / datasets** we could reuse?

### imgproxy ground truth

The actual autoquality search loop is **Pro/closed-source** — not in the open repo. What's public:

- The `size`/`dssim` methods re-encode the **full result** several times (docs warn: *"requires an image to be saved several times… prefer quality over speed"*).
- The `ml` method feeds a **256×256×3 downsized RGB image into a trained ONNX net** that predicts quality, then (unless `allowed_error: 1`) **refines with the real DSSIM search**. That imgproxy needed a *learned* model rather than naive metric-transfer is itself the signal that a downsized proxy alone doesn't transfer cleanly.

### Q1 — downsized proxy: fast, but doesn't transfer accurately

Measured JPEG size + a luminance-SSIM DSSIM on `high_freq.jpg` (1600×1200, deliberately worst-case high-frequency content), full res vs proxies:

```
JPEG SIZE        q40      q50      q60      q70      q75      q80      q85      q90
1/1 1600x1200  221841   246463   266427   313511   330425   356741   434144   522559
1/2 800x600     78167    86071    93934   106227   113183   125113   140248   173690
1/4 400x300     27083    29811    32658    36637    38991    42718    47668    57631

DSSIM            q40      q50      q60      q70      q75      q80      q85      q90
1/1            0.0052   0.0040   0.0030   0.0023   0.0018   0.0015   0.0011   0.0003
1/2            0.0028   0.0021   0.0017   0.0012   0.0010   0.0008   0.0006   0.0004
1/4            0.0024   0.0018   0.0014   0.0009   0.0008   0.0006   0.0004   0.0002
1/8            0.0095   0.0067   0.0049   0.0032   0.0024   0.0016   0.0010   0.0005
```

- **Speed: large win.** Encode cost is ~linear in pixels; a 1/4 proxy encodes 1/16 the pixels. A 5-iteration search drops from ~5 full encodes to ~5 tiny encodes + 1 final full encode ≈ the cost of a single full encode.
- **Accuracy: no, not by direct metric-transfer.** Neither target is scale-invariant:
  - **`size`** — proxy bytes are unrelated to full-res bytes, *and the curve shape changes* (q90/q40 size ratio: 2.36× full vs 2.13× at 1/4). No fixed multiplier recovers full-res size; the size method **must measure on the full result**.
  - **`dssim`** — at q75 the 1/2 and 1/4 proxies report *lower* damage than full res (downsizing low-passes the high frequencies JPEG quantizes hardest), so a proxy search to hit 0.002 picks ~q55, which on the full image lands ~0.0035 (~75% past target). Worse, it's **non-monotonic**: 1/8 *reverses* (aliasing). A bias whose sign flips with scale can't be fixed by a constant offset — exactly why imgproxy used a trained net.

Caveat: DSSIM here is a standard SSIM-window approximation (σ=1.5 luminance), not imgproxy's exact dssim — absolute values differ, but the **cross-scale bias direction** is the robust finding.

**Implication:** the accuracy-safe use of a proxy is to **seed the initial quality/bracket**, then **converge on full-res** — not to search entirely on the proxy.

### Q2 — prebuilt models / methods / datasets

- **imgproxy's own models: not reusable.** Bundled only in the paid Pro `-ml` Docker image, proprietary; methodology + training set unpublished ("contact us"). Only the I/O contract is public: ONNX, 256×256×3 RGB in [0,1] → 1 scalar in [0,1].
- **Open ecosystem solved it without a quality-prediction CNN** — via *metric-targeted encode search* with fast perceptual metrics:
  - **SSIMULACRA2** / **butteraugli** (JPEG XL effort; SSIMULACRA2 is multi-scale, built for compression artifacts). Fast/GPU impls exist ([Vship](https://github.com/Line-fr/Vship), [fast SSIMULACRA2](https://www.researchgate.net/publication/401347718_Fast_implementation_of_SSIMULACRA2_for_image_quality_assessment)).
  - **[jpegli-rs](https://crates.io/crates/jpegli-rs)** has built-in `Quality::ApproxSsim2` / `Quality::ApproxButteraugli` targets — essentially our `dssim` method with a better metric.
  - **[Auto-Target-Encoder](https://github.com/Snickrr/Auto-Target-Encoder)**: a RandomForest predicts the CRF to hit a target metric, then refines — the *seed-then-converge* pattern, lightweight (forest, not CNN).
  - [Av1an target-quality](https://rust-av.github.io/Av1an/Cli/target_quality.html), [Guetzli](https://arxiv.org/pdf/1703.04421) (closed-loop butteraugli), Cloudinary `q_auto` (trial encodes scored by L\*a\*b\* multi-scale DSSIM).
- **If we ever train our own**: self-supervised QF-predictor pattern ([ref](https://arxiv.org/html/2405.02208)); IQA datasets with JPEG distortions — **KADID-10k**, **TID2013**, **CLIC2021**, **PIPAL**. Predicts a *score*, not "the QF to hit target X", so still pairs with a search. Large lift (ONNX inference dep) for marginal gain.

### Recommendation for this issue

1. Ship the `size` + `dssim` search on the **full-res result** first (imgproxy parity; shared encode-search machinery with #83 `max_bytes`).
2. Treat **proxy** purely as a future speed optimization in its safe form: proxy to **seed** the bracket, converge on full-res. Never search entirely on the proxy for `size`/`dssim`.
3. Treat **`ml`** as a much later optional stub — there is **no license-clean drop-in model today**. If pursued, the open path is metric-targeted search with **SSIMULACRA2/butteraugli** (better scale behavior than single-scale DSSIM, and cheap per-iteration), optionally with a small learned *seed*.

Reject unsupported `method` values (anything beyond what we implement) before side effects, per the issue's realization notes.
